---
title: "GenomicSelection"
author: "Camilo E. SÃ¡nchez & Danilo Moreta"
date: "Cali, CO | 26 February 2024"
output:
  html_document:
    fig_caption: yes
    fig_height: 7
    fig_width: 5
    highlight: haddock
    number_sections: yes
    theme: cosmo
    toc: yes
    toc_float:
      collapsed: no
      toc_depth: 4
  pdf_document:
    toc: yes
  word_document:
    toc: yes
editor_options: 
  chunk_output_type: console
---

# 1: Define inputs

```{r}
dir <- 
  'D:/OneDrive - CGIAR/00_CassavaBioinformaticsPlatform/07_GenomicSelection/03_DaRTSeq/'
geno_file <- 'Report_DCas23-8685_SNP.csv'
pheno_file <- 'GS_2022_first_cohort.xlsx'
```

# 2: Setting working directory

```{r}
getwd()
setwd(dir)
getwd()
```

# 3: Install packages

```{r}
if (!require(data.table)) install.packages(data.table)
if (!require(readxl)) install.packages(readxl)
if (!require(tidyverse)) install.packages(tidyverse)
if (!require(rrBLUP)) install.packages(rrBLUP)
```

# 4: Load libraries

```{r}
library(data.table)
library(readxl)
library(tidyverse)
library(rrBLUP)
```

# 5: Read files

```{r}
# Read geno data
geno <- as.data.frame(t(fread(geno_file))) %>%
  slice(-2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -13, -14, -15, -16) %>%
  select(-1, -2, -3, -4)

# Read pheno data, filter by group2, delete unnecessary columns and rename the accessions
pheno <- read_excel(pheno_file) %>%
  dplyr::filter(group == 'group2') %>%
  select(-1, -2) %>%
  rename(Genotype = geno_accession_name)

# Display the first few rows of each data frame to check
head(geno)
head(pheno)
```

# 6: Modify geno file

```{r}
# Replace a value inside the data frame
geno[1,1] <- "Plates"

# Change column and row names
colnames(geno) <- c(geno[1,])
rownames(geno) <- c(1:nrow(geno))

# Delete first row, filter by plate, delete firts column and chage markers names
geno <- geno %>% slice(-1) %>%
  dplyr::filter(Plates == 'Group2_Plate01' | Plates == 'Group2_Plate02' |
                  Plates == 'Group2_Plate03' | Plates == 'Group2_Plate04') %>%
  select(-1) %>%
  rename(Genotype = MarkerName)
```

# 7: Joint accession names

```{r}
# Extract genotypes names of geno and pheno file
geno_names <- as.data.frame(geno$Genotype)
pheno_names <- as.data.frame(pheno$Genotype)

# Separate the names
geno_names <- geno_names %>% 
  separate(`geno$Genotype`, into = c("Date", "Trial", "Genotype"), sep = "_") %>%
  select(Genotype)

# Separate the names
pheno_names <- pheno_names %>% 
  separate(`pheno$Genotype`, into = c("Date", "Trial", "Genotype"), sep = "_")

pheno_names <- replace(pheno_names, is.na(pheno_names), '-')

pheno_names <- pheno_names %>%
  mutate(Genotype = ifelse(Genotype == '-', paste0(Date), Genotype)) %>%
  select(Genotype)

# Find duplicate values between pheno and geno files
duplicate_names <- intersect(geno_names$Genotype, pheno_names$Genotype)

# Replace genotypes names
geno$Genotype <- geno_names$Genotype
pheno$Genotype <- pheno_names$Genotype

#Filter unused genotypes
# Geno file
geno <- geno %>%
  dplyr::filter(Genotype %in% duplicate_names)

# Pheno file
pheno <- pheno %>%
  dplyr::filter(Genotype %in% duplicate_names)
```

# 8: Messages

```{r}
# Print the number of samples and SNPs in genomic data files
message(paste0("Number of samples in geno file: ", nrow(geno) , "\n",
               "Number of samples in pheno file: ", nrow(pheno), "\n\n",
               "WARNING: The samples in both files must be the same"))

# Print the number of SNPs and traits to analyze
message(paste0("Number of SNPs: ", ncol(geno) - 1 , "\n",
               "Number of traits: ", ncol(pheno) - 1))
```

# 9: Re-formatting SNP values

```{r}
geno[geno == 0] <- -1
geno[geno == 1] <- 0
geno[geno == 2] <- 1
geno[geno == '-'] <- NA
```

# 10: Source function: GS

```{r}
# Get the function
source('D:/OneDrive - CGIAR/00_CassavaBioinformaticsPlatform/00_Basics/04_GitHub/CassBio.GenomicSelection/R_code/GenomicSelection.R')
```

# 11: Cross validation for single trait

```{r}
# Define inputs
imp_method <- 'mean'
GS_method <- 'REML'
cv <- 100
prop <- 0.8
single_trait <- 'yield_ha'

# Create an empty data frame to storage results
trait_df <- data.frame(matrix(ncol = 6, nrow = 0))

# Create and set a directory to store the results
setwd(dir)
dir.create(paste0('GS_Results_SingleTrait_', single_trait, '_', cv, '_', prop))
setwd(paste0('GS_Results_SingleTrait_', single_trait, '_', cv, '_', prop))

# Define the time of starting
start_time <- Sys.time()

# Perform the function using CV for a single trait
for (k in 1:cv){
    
    # Perform the function
    result <- GS(single_trait, pheno, geno, prop, k, imp_method, GS_method)
    
    # Store the results
    trait_df <- rbind(trait_df, data.frame(result))
    
    # Progress bar
    cat('\r', k, ' files processed |', rep('=', k / 10), 
        ifelse(k == nrow(cv), '|\n', '>'), sep = '')
    
}

# Save the results
write.csv(trait_df, paste0(single_trait, '_', cv, '_', prop, '.csv'), row.names = F)

# Calculate ending time
end_time <- Sys.time()

# Print running time
message(paste0('Function ends of running. Time taken: ', (end_time - start_time)))
```

# 12: Cross validation for all traits

```{r}
# Define inputs
imp_method <- 'mean'
GS_method <- 'REML'
cv <- 50
prop <- 0.8

# Create an empty data frame to storage results
trait_df <- data.frame(matrix(ncol = 6, nrow = 0))

# Create and set a directory to store the results
setwd(dir)
dir.create(paste0('GS_Results_AllTraits_', cv, '_', prop))
setwd(paste0('GS_Results_AllTraits_', cv, '_', prop))

# Define the time of starting
start_time <- Sys.time()

# Perform the function complete using all traits and CV
for (single_trait in colnames(pheno[,-1])){
  
  # Informative message
  message(paste0('Working on trait: ', single_trait, '\n'))
    
  # Cross validation for each trait
  for (k in 1:cv){
      
    # Perform the function
    result <- GS(single_trait, pheno, geno, prop, k, imp_method, GS_method)
    
    # Store the results
    trait_df <- rbind(trait_df, data.frame(result))
      
    # Progress bar
    cat('\r', k, ' files processed |', rep('=', k / 10), 
        ifelse(k == nrow(cv), '|\n', '>'), sep = '')
      
    }
    
  # Informative message
  message(paste0('\n', 'Done on trait: ', single_trait, '\n'))
    
}

# Save the results
write.csv(trait_df, paste0('GS_Results_AllTraits_', cv, '_', prop, '.csv'), row.names = F)

# Calculate ending time
end_time <- Sys.time()

# Print running time
message(paste0('Function ends of running. Time taken: ', (end_time - start_time)))
```

# 13: Cross validation using all traits and different prop values

```{r}
# Define inputs
imp_method <- 'mean'
GS_method <- 'REML'
cv <- 20
props <- c(0.9, 0.8, 0.7, 0.6, 0.5)

# Create an empty data frame to storage results
trait_df <- data.frame(matrix(ncol = 6, nrow = 0))

# Create and set a directory to store the results
setwd(dir)
dir.create(paste0('GS_Results_AllTraits_DiffProp_', cv))
setwd(paste0('GS_Results_AllTraits_DiffProp_', cv))

# Define starting point
start_time <- Sys.time()

# Perform the function complete using all traits, CV and different prop
for (single_trait in colnames(pheno[,-1])){
  
  # Informative message
  message(paste0('Working on trait: ', single_trait, '\n'))
    
  # Iterate over the different prop values
  for (prop in props){
    
    # Informative message
    message(paste0('% of validation set: ', 1 - prop, '\n'))
    
    # Cross validation for each trait
    for (k in 1:cv){
      
      # Perform the function
      result <- GS(single_trait, pheno, geno, prop, k, imp_method, GS_method)
      
      # Store the results
      trait_df <- rbind(trait_df, data.frame(result))
      
      # Progress bar
      cat('\r', k, ' files processed |', rep('=', k / 10), 
          ifelse(k == nrow(cv), '|\n', '>'), sep = '')
      
    }
    
    # Informative message
    message(paste0('\n', 'Done on validation set of: ', 1 - prop, '%', '\n'))
  
  }
  
  # Informative message
  message(paste0('\n', 'Done on trait: ', single_trait, '\n'))
  
}

# Save the results
write.csv(trait_df, paste0('GS_Results_AllTraits_DiffProp_', cv, '.csv'), row.names = F)

# Calculate ending time
end_time <- Sys.time()

# Print running time
message(paste0('Function ends of running. Time taken: ', (end_time - start_time)))
```

# 14: Source function: Plots

```{r}
# Source plotting function
source('D:/OneDrive - CGIAR/00_CassavaBioinformaticsPlatform/00_Basics/04_GitHub/CassBio.GenomicSelection/R_code/AccuracyPlot.R')

# Execute function
AccuracyPlot(trait_df, type = 'single') ############

```

# 15: Source function: Genomic Estimated Breeding Values (GEBVs)

```{r}
# Source plotting function
source('D:/OneDrive - CGIAR/00_CassavaBioinformaticsPlatform/00_Basics/04_GitHub/CassBio.GenomicSelection/R_code/GEBVs.R')
```

# 16: GEBVs for all traits

```{r}
# Define inputs
imp_method <- 'mean'
GS_method <- 'REML'
prop <- 0.8

# Create an empty data frame to storage results
GEBVs_df <- data.frame(matrix(ncol = 6, nrow = 0))

# Create and set a directory to store the results
setwd(dir)
dir.create(paste0('GS_GEBVs_Results_', prop))
setwd(paste0('GS_GEBVs_Results_', prop))

# Define starting point
start_time <- Sys.time()

# Perform the function complete using all traits and CV
for (single_trait in colnames(pheno[,-1])){
  
  # Informative message
  message(paste0('Working on trait: ', single_trait, '\n'))
  
  # Perform the function
  result <- GEBVs(single_trait, pheno, geno, prop, imp_method, GS_method)
    
  # Store the results
  GEBVs_df <- rbind(GEBVs_df, data.frame(result))
  
  # Informative message
  message(paste0('\n', 'Done on trait: ', single_trait, '\n'))s
    
}

# Save the results
write.csv(GEBVs_df, paste0('GS_GEBVs_Results_', prop, '.csv'), row.names = F)

# Calculate ending time
end_time <- Sys.time()

# Print running time
message(paste0('Function ends of running. Time taken: ', (end_time - start_time)))

```
